includes/class-amp-service-worker.php

```diff
@@ -0,0 +1,316 @@
+<?php
+/**
+ * AMP Service Workers.
+ *
+ * @package AMP
+ * @since 1.1
+ */
+
+/**
+ * Class AMP_Service_Worker.
+ */
+class AMP_Service_Worker {
+
+	/**
+	 * Query var that is used to signal a request to install the service worker in an iframe.
+	 *
+	 * @link https://www.ampproject.org/docs/reference/components/amp-install-serviceworker#data-iframe-src-(optional)
+	 */
+	const INSTALL_SERVICE_WORKER_IFRAME_QUERY_VAR = 'amp_install_service_worker_iframe';
+
+	/**
+	 * Init.
+	 */
+	public static function init() {
+		if ( ! class_exists( 'WP_Service_Workers' ) ) {
+			return;
+		}
+
+		$theme_support = AMP_Theme_Support::get_theme_support_args();
+		if ( isset( $theme_support['service_worker'] ) && false === $theme_support['service_worker'] ) {
+			return;
+		}
+
+		add_filter( 'query_vars', array( __CLASS__, 'add_query_var' ) );
+		add_action( 'parse_request', array( __CLASS__, 'handle_service_worker_iframe_install' ) );
+		add_action( 'wp', array( __CLASS__, 'add_install_hooks' ) );
+
+		if ( true === $theme_support['service_worker'] || ! empty( $theme_support['service_worker']['cdn_script_caching'] ) ) {
+			add_action( 'wp_front_service_worker', array( __CLASS__, 'add_cdn_script_caching' ) );
+		}
+		if ( true === $theme_support['service_worker'] || ! empty( $theme_support['service_worker']['image_caching'] ) ) {
+			add_action( 'wp_front_service_worker', array( __CLASS__, 'add_image_caching' ) );
+		}
+		if ( true === $theme_support['service_worker'] || ! empty( $theme_support['service_worker']['google_fonts_caching'] ) ) {
+			add_action( 'wp_front_service_worker', array( __CLASS__, 'add_google_fonts_caching' ) );
+		}
+	}
+
+	/**
+	 * Add query var for iframe service worker request.
+	 *
+	 * @param array $vars Query vars.
+	 * @return array Amended query vars.
+	 */
+	public static function add_query_var( $vars ) {
+		$vars[] = self::INSTALL_SERVICE_WORKER_IFRAME_QUERY_VAR;
+		return $vars;
+	}
+
+	/**
+	 * Add runtime caching for scripts loaded from the AMP CDN with a stale-while-revalidate strategy.
+	 *
+	 * @link https://github.com/ampproject/amp-by-example/blob/4593af61609898043302a101826ddafe7206bfd9/boilerplate-generator/templates/files/serviceworkerJs.js
+	 *
+	 * @param WP_Service_Worker_Scripts $service_workers Service worker registry.
+	 */
+	public static function add_cdn_script_caching( $service_workers ) {
+		if ( ! ( $service_workers instanceof WP_Service_Worker_Scripts ) ) {
+			_doing_it_wrong( __METHOD__, esc_html__( 'Expected argument to be WP_Service_Worker_Cache_Registry.', 'amp' ), '1.1' );
```

Comment:
I believe the reason for this change is because the argument was different in previous iteration of the plugin, so this is to prevent a fatal error from happening.

Reply:
Ah, good catch. I think in this case we should omit the `_doing_it_wrong()` though and just return. This method is not manually called by anybody, so neither the user nor the developer is really "at fault" for this happening. It will be the case if someone uses this version of the AMP plugin with version 0.1 of the PWA plugin.

Reply:
@felixarntz then what about just changing the message to say: "Please upgrade to PWA plugin v0.2".

Reply:
Like so:

```diff
diff --git a/includes/class-amp-service-worker.php b/includes/class-amp-service-worker.php
index 37dbc0e1..73a2280e 100644
--- a/includes/class-amp-service-worker.php
+++ b/includes/class-amp-service-worker.php
@@ -83,7 +83,7 @@ class AMP_Service_Worker {
 	 */
 	public static function add_cdn_script_caching( $service_workers ) {
 		if ( ! ( $service_workers instanceof WP_Service_Worker_Scripts ) ) {
-			_doing_it_wrong( __METHOD__, esc_html__( 'Expected argument to be WP_Service_Worker_Cache_Registry.', 'amp' ), '1.1' );
+			_doing_it_wrong( __METHOD__, esc_html__( 'Please update to PWA v0.2. Expected argument to be WP_Service_Worker_Cache_Registry.', 'amp' ), '1.1' );
 			return;
 		}
 
@@ -125,7 +125,7 @@ class AMP_Service_Worker {
 	 */
 	public static function add_image_caching( $service_workers ) {
 		if ( ! ( $service_workers instanceof WP_Service_Worker_Scripts ) ) {
-			_doing_it_wrong( __METHOD__, esc_html__( 'Expected argument to be WP_Service_Worker_Scripts.', 'amp' ), '1.1' );
+			_doing_it_wrong( __METHOD__, esc_html__( 'Please update to PWA v0.2. Expected argument to be WP_Service_Worker_Scripts.', 'amp' ), '1.1' );
 			return;
 		}
 
@@ -157,6 +157,10 @@ class AMP_Service_Worker {
 	 * @param WP_Service_Worker_Scripts $service_workers Service workers.
 	 */
 	public static function add_google_fonts_caching( $service_workers ) {
+		if ( ! ( $service_workers instanceof WP_Service_Worker_Scripts ) ) {
+			_doing_it_wrong( __METHOD__, esc_html__( 'Please update to PWA v0.2. Expected argument to be WP_Service_Worker_Scripts.', 'amp' ), '1.1' );
+			return;
+		}
 
 		// The PWA plugin also automatically adds runtime caching for Google Fonts when WP_SERVICE_WORKER_INTEGRATIONS_ENABLED is set.
 		if ( class_exists( 'WP_Service_Worker_Fonts_Integration' ) ) {
```

It think it's important to keep _some_ kind of notice, as otherwise it just silently fails.